### PR TITLE
HPCC-28079 Revisit roxie cache warming

### PR DESCRIPTION
**Pull request by richardkchapman**
Some additional notes and thoughts.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: devdoc/roxie.md [hunk 1]
```diff
@@ -189,4 +189,45 @@ Having first compiled a suitable bit of ECL into a.out. I have found a snippet l
...
+
+Thirdly, need to check that the prewarm is complete before adding a new agent to the topology. This is especially relevant if we make
+any change to do the prewarm asynchronously.
+
+Fourthly, there are potential race conditions when reading/writing the file containing cache information, since this file may be written
+by any agent operating on the same channel, at any time.
+
+Fifthly, how is the amount of information tracked decided? It should be at least related to the amount of memory available to the linux
+page cache, but that's not a completely trivial thing to calculate. Should we restrict to the most recent N when outputting, where N is
+calculated from, for example /proc/meminfo's Active(file) value? Unfortunately on containerized sytems that reflects tht host, but perhaps
```
Comment (ghalliday) [inline devdoc/roxie.md:222]: typo: that/the?

File: devdoc/roxie.md [hunk 2]
```diff
@@ -189,4 +189,45 @@ Having first compiled a suitable bit of ECL into a.out. I have found a snippet l
...
+/sys/fs/cgroup/memory.stat can be used instead? 
+
+When deciding how much to track, we can pick an upper limit from the pod's memory limit. This could be read from /sys/fs/cgroup/memory.max
+though we currently read from the config file instead. We should probably (a) subtract the roxiemem size from that and (b) think about
+a value that will work on bare-metal and fusion too. However, because we don't dedup the entries in the circular buffer used for tracking
+hot pages unti lthe info is flushed, the appropriate size is not really the same as the memory size.
```
Comment (ghalliday) [inline devdoc/roxie.md:228]: typo: until the

File: devdoc/roxie.md [hunk 3]
```diff
@@ -189,4 +189,45 @@ Having first compiled a suitable bit of ECL into a.out. I have found a snippet l
...
+
+There are some questions outstanding about how this operates that may require empirical testing to answer. Firstly, how does this
+interact with volumes mounted via k8s pvc's, and in particular with cloud billing systems that charge per read. Will the reads
+that are done to warm the cache be done in large chunks, or will they happen one linux page at a time? The code at the Roxie level
+operates by memory-mapping the file then touching a byte within each linux page that we want to be "warm", but does the linux paging
+subsystem fetch larger blocks? Do huge pages play a part here?
+
+Secondly, the prewarm is actually done by a child process (ccdcache), but the parent process is blocked while it happens. It would
+probably make sense to at allow at least some of the other startup operations of the parent process to proceed in parallel. I don't
+recall why it was done via a child process - I suspect that it was coded to allow the prewarm to be done manually (e.g. in the bare-
```
Comment (ghalliday) [inline devdoc/roxie.md:210]: My memory is that it is something to do with catching segment faults that can occur, and making sure they don't terminate the roxie process.

**Review by ghalliday — COMMENTED** [review]
@richardkchapman a couple of typos

File: devdoc/roxie.md (hunk 3 above)
Comment (richardkchapman) [reply devdoc/roxie.md:210]: Ah, that would make sense. I will update the doc accordingly.

**Comment by richardkchapman** [comment]
Updated

**Review by ghalliday — APPROVED** [review]
@richardkchapman please squash

**Comment by richardkchapman** [comment]
@ghalliday Done

**Pull request merged**